File: frontend/src/api/quiz.ts
import { apiClient } from "../lib/apiClient";
import { Quiz, QuizAttempt } from "../types";

export type QuizWithQuestions = Quiz & {
  questions: {
    id: string;
    quiz_id: string;
    question_text: string;
    options: string[];
    correct_option_index: number;
    reasoning: string;
  }[];
  status: "active" | "completed";
};

export async function generateQuiz(
  boardId: string,
  maxQuestions: number
): Promise<QuizWithQuestions> {
  const response = await apiClient.post<QuizWithQuestions>(
    `/boards/${boardId}/quizzes/generate`,
    { max_questions: maxQuestions }
  );
  return response.data;
}

export async function getActiveQuiz(boardId: string): Promise<QuizWithQuestions | null> {
  const response = await apiClient.get<QuizWithQuestions | null>(`/boards/${boardId}/quizzes/active`);
  return response.data;
}

export async function getQuiz(quizId: string): Promise<QuizWithQuestions> {
  const response = await apiClient.get<QuizWithQuestions>(`/quizzes/${quizId}`);
  return response.data;
}

export async function submitAttempt(
  quizId: string,
  userAnswers: Record<string, number>
): Promise<QuizAttempt> {
  const response = await apiClient.post<QuizAttempt>(
    `/quizzes/${quizId}/attempts`,
    { user_answers: userAnswers }
  );
  return response.data;
}

File: backend/app/models/quiz.py
import uuid
from datetime import datetime
from sqlalchemy import String, ForeignKey, DateTime, func, Text, Integer, JSON
from sqlalchemy.orm import Mapped, mapped_column
from app.database import Base

class Quiz(Base):
    __tablename__ = "quizzes"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: uuid.uuid4().hex)
    board_id: Mapped[str] = mapped_column(String(36), ForeignKey("boards.id", ondelete="CASCADE"), nullable=False)
    max_questions: Mapped[int] = mapped_column(Integer, nullable=False)
    status: Mapped[str] = mapped_column(String(20), default="active", nullable=False)  # "active" or "completed"
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), onupdate=func.now())


class QuizQuestion(Base):
    __tablename__ = "quiz_questions"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: uuid.uuid4().hex)
    quiz_id: Mapped[str] = mapped_column(String(36), ForeignKey("quizzes.id", ondelete="CASCADE"), nullable=False)
    question_text: Mapped[str] = mapped_column(Text, nullable=False)
    options: Mapped[list[str]] = mapped_column(JSON, nullable=False)  # JSON list of strings
    correct_option_index: Mapped[int] = mapped_column(Integer, nullable=False)
    reasoning: Mapped[str] = mapped_column(Text, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())


class QuizAttempt(Base):
    __tablename__ = "quiz_attempts"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: uuid.uuid4().hex)
    quiz_id: Mapped[str] = mapped_column(String(36), ForeignKey("quizzes.id", ondelete="CASCADE"), nullable=False)
    user_answers: Mapped[dict[str, int]] = mapped_column(JSON, nullable=False)  # JSON question_id -> chosen index
    score: Mapped[int] = mapped_column(Integer, nullable=False)
    completed_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
